sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -88,5 +85,158 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+            private DateTimeOffset _refreshOn;
+            private DateTimeOffset _expiresOn;
+
+            public AccessTokenCache(TimeSpan tokenRefreshOffset)
+            {
+                _tokenRefreshOffset = tokenRefreshOffset;
+            }
+
+            public async ValueTask<(string? headerValue, bool scheduleRefreshToken)> GetHeaderValueAsync(bool async, CancellationToken cancellationToken)
+            {
+                (string? headerValue, Task<string>? pendingTask, bool scheduleRefreshToken) = GetHeaderValue();
+                if (pendingTask == null)
+                {
+                    return (headerValue, scheduleRefreshToken);
+                }
+
+                if (async)
+                {
+                    return (await pendingTask.AwaitWithCancellation(cancellationToken), false);
+                }
+
+                try
+                {
+                    pendingTask.Wait(cancellationToken);
+                }
+                catch (AggregateException) { } // ignore exception here to rethrow it with EnsureCompleted
+
+                return (pendingTask.EnsureCompleted(), false);
+            }
+
+            public string SaveToken(AccessToken token)
+            {
+                if (string.IsNullOrEmpty(token.Token))
+                {
+                    var exception = new InvalidOperationException($"{nameof(TokenCredential)}.{nameof(TokenCredential.GetToken)} has failed with unknown error.");
+                    ResetTokenIfExpired(exception);
+                    throw exception;
+                }
+
+                string headerValue;
+                TaskCompletionSource<string>? pendingTcs;
+
+                lock (_syncObj)
+                {
+                    _headerValue = "Bearer " + token.Token;
+                    _refreshOn = token.ExpiresOn - _tokenRefreshOffset;
+                    _expiresOn = token.ExpiresOn;
+                    _tokenState = TokenState.Valid;
+
+                     pendingTcs = _pendingTcs;
+                    headerValue = _headerValue;
+                    _pendingTcs = null;
+                }
+
+                pendingTcs?.SetResult(headerValue);
+                return headerValue;
+            }
+
+            public void ResetTokenIfExpired(Exception exception)
+            {
+                TaskCompletionSource<string>? pendingTcs = default;
+
+                lock (_syncObj)
+                {
+                    if (DateTimeOffset.UtcNow >= _expiresOn)
+                    {
+                        _headerValue = default;
+                        _refreshOn = default;
+                        _expiresOn = default;
+                        _tokenState = TokenState.Invalid;
+
+                        pendingTcs = _pendingTcs;
+                        _pendingTcs = null;
+                    }
+                    else if (_tokenState == TokenState.AboutToExpire)
+                    {
+                        _tokenState = TokenState.Valid;
```

Comment:
Why does an Exception when the state is `AboutToExpire` result in setting the state to `Valid`?

Reply:
Since token hasn't expired, we want to return back to `Valid` state so that next call can make another attempt

Reply:
I think we need a detailed comment somewhere on this class which describes the meaning of each state and how they are used.

Reply:
Refactored this piece. State field is replaced with `_isGettingToken` which is `true` while `GetToken` is running.

Reply:
States are removed in new implementation